src/lib/snapper/snapshot.rb

```diff
@@ -0,0 +1,199 @@
+# encoding: utf-8
+
+# ------------------------------------------------------------------------------
+# Copyright (c) 2016 SUSE LLC. All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it under
+# the terms of version 2 of the GNU General Public License as published by the
+# Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, contact Novell, Inc.
+#
+# To contact Novell about this file by physical or electronic mail, you may find
+# current contact information at www.novell.com.
+# ------------------------------------------------------------------------------
+
+require "yast"
+require "snapper/snapshot_dbus"
+
+module Yast
+  ##
+  # This class represents the base class for Snapshots mostly based on Snapper
+  # ones.
+  class Snapshot
+    include Yast::Logger
+
+    attr_reader :number, :pre_number, :timestamp, :user
+
+    attr_accessor :description, :cleanup, :user_data, :config, :communication
+
+    class << self
+      # @return [Hash] Mapping string with available subclass
+      def types
+        { "single" => SingleSnapshot, "post" => PostSnapshot, "pre" => PreSnapshot }
+      end
+
+      # @return default strategy responsable of talk with the underlying system
+      def default_communication(context = nil)
+        SnapshotDBus.new(context)
+      end
+
+      # @return [Array] of Snapshots
+      def all(configs = nil, communication = default_communication)
```

Comment:
I really prefer named parameters instead of common ones with default, this way it is hard to change only communication without configs.